barretenberg/cpp/src/barretenberg/execution_trace/execution_trace.cpp

```diff
@@ -55,12 +55,61 @@ template <class Flavor>
     Builder& builder, typename Flavor::ProvingKey& proving_key, bool is_structured)
 {
-    ZoneScopedN("construct_trace_data");
-    TraceData trace_data{ builder, proving_key };
-
     // Complete the public inputs execution trace block from builder.public_inputs
     populate_public_inputs_block(builder);
 
+    ZoneScopedN("construct_trace_data");
+    // Allocate the wires and selectors polynomials
+    if constexpr (IsHonkFlavor<Flavor>) {
+        for (auto& wire : proving_key.polynomials.get_wires()) {
+            wire = Polynomial::shiftable(proving_key.circuit_size);
+        }
+        // Define selectors over the block they are isolated to
+        uint32_t offset = Flavor::has_zero_row ? 1 : 0;
+        if constexpr (IsGoblinFlavor<Flavor>) {
+            offset += builder.blocks.ecc_op.get_fixed_size(is_structured);
+        }
+        offset += builder.blocks.pub_inputs.get_fixed_size(is_structured);
+
+        proving_key.polynomials.q_arith =
+            Polynomial(proving_key.circuit_size - offset, proving_key.circuit_size, offset);
+        offset += builder.blocks.arithmetic.get_fixed_size(is_structured);
+
+        proving_key.polynomials.q_delta_range =
+            Polynomial(builder.blocks.delta_range.size(), proving_key.circuit_size, offset);
+        offset += builder.blocks.delta_range.get_fixed_size(is_structured);
+
+        proving_key.polynomials.q_elliptic =
+            Polynomial(builder.blocks.elliptic.size(), proving_key.circuit_size, offset);
+        offset += builder.blocks.elliptic.get_fixed_size(is_structured);
+
+        proving_key.polynomials.q_aux = Polynomial(builder.blocks.aux.size(), proving_key.circuit_size, offset);
+        offset += builder.blocks.aux.get_fixed_size(is_structured);
+
+        proving_key.polynomials.q_lookup = Polynomial(builder.blocks.lookup.size(), proving_key.circuit_size, offset);
+        offset += builder.blocks.lookup.get_fixed_size(is_structured);
+
+        if constexpr (HasDataBus<Flavor>) {
+            proving_key.polynomials.q_busread =
+                Polynomial(builder.blocks.busread.size(), proving_key.circuit_size, offset);
+            offset += builder.blocks.busread.get_fixed_size(is_structured);
+        }
+
+        proving_key.polynomials.q_poseidon2_external =
+            Polynomial(builder.blocks.poseidon2_external.size(), proving_key.circuit_size, offset);
+
+        offset += builder.blocks.poseidon2_external.get_fixed_size(is_structured);
+        proving_key.polynomials.q_poseidon2_internal =
+            Polynomial(builder.blocks.poseidon2_internal.size(), proving_key.circuit_size, offset);
+        offset += builder.blocks.poseidon2_internal.get_fixed_size(is_structured);
+        // Set the other selector polynomials to full size
+        for (auto& selector : proving_key.polynomials.get_non_gate_selectors()) {
+            selector = Polynomial(proving_key.circuit_size);
```

Comment:
these along with the wires, sigmas, ids, and z_perm could be restricted to the actual circuit size and not the dyadic circuit_size.

Reply:
or potentially more if we don't even use some blocks (which means getting rid of the ensure_nonzero stuff)